source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -506,16 +616,39 @@ int ContextImpl::ignoreCertificateExpirationCallback(int ok, X509_STORE_CTX* ctx
   return ok;
 }
 
-int ContextImpl::verifyCallback(X509_STORE_CTX* store_ctx, void* arg) {
-  ContextImpl* impl = reinterpret_cast<ContextImpl*>(arg);
-  SSL* ssl = reinterpret_cast<SSL*>(
-      X509_STORE_CTX_get_ex_data(store_ctx, SSL_get_ex_data_X509_STORE_CTX_idx()));
+ssl_verify_result_t ContextImpl::verifyCallback(SSL* ssl, uint8_t* out_alert) {
+  ContextImpl* impl = reinterpret_cast<ContextImpl*>(SSL_CTX_get_app_data(SSL_get_SSL_CTX(ssl)));
   Envoy::Ssl::SslExtendedSocketInfo* sslExtendedInfo =
       reinterpret_cast<Envoy::Ssl::SslExtendedSocketInfo*>(
           SSL_get_ex_data(ssl, ContextImpl::sslExtendedSocketInfoIndex()));
 
+  const STACK_OF(CRYPTO_BUFFER)* buffers = SSL_get0_peer_certificates(ssl);
+  bssl::UniquePtr<X509> cert = nullptr;
+  bssl::UniquePtr<STACK_OF(X509)> chain(sk_X509_new(nullptr));
+  for (size_t i = 0; i < sk_CRYPTO_BUFFER_num(buffers); ++i) {
+    bssl::UniquePtr<X509> x509(X509_parse_from_buffer(sk_CRYPTO_BUFFER_value(buffers, i)));
+    if (!x509) {
+      continue;
+    }
+    if (!cert) {
+      cert = std::move(x509);
+    } else {
+      bool rc = bssl::PushToStack(chain.get(), std::move(x509));
+      RELEASE_ASSERT(rc == true, Utility::getLastCryptoError().value_or(""));
+    }
+  }
+
   if (impl->verify_trusted_ca_) {
-    int ret = X509_verify_cert(store_ctx);
+    bssl::ScopedX509_STORE_CTX store_ctx;
+    if (!X509_STORE_CTX_init(store_ctx.get(), impl->tls_contexts_[0].x509_store_.get(), cert.get(),
+                             chain.get()) ||
+        !X509_STORE_CTX_set_default(store_ctx.get(),
+                                    SSL_is_server(ssl) ? "ssl_client" : "ssl_server")) {
```

Comment:
From the docs (https://www.openssl.org/docs/manmaster/man3/X509_STORE_CTX_new.html), it looks like these strings are used to lookup an entry. Is there a header or API function in boringssl to get these values from?

Reply:
Not that I could find.